packages/gatsby-plugin-image/src/transforms/gatsby-plugin-image.js

```diff
@@ -0,0 +1,302 @@
+      MemberExpression(path) {
+        if (
+          (path.node?.property?.name === `fixed` ||
+            path.node?.property?.name === `fluid`) &&
+          path.node?.object?.property?.name === `childImageSharp`
+        ) {
+          const updatedExpression = t.memberExpression(
+            path.node.object,
+            t.identifier(`gatsbyImageData`)
+          )
+          path.replaceWith(updatedExpression)
+        }
+      },
+      JSXOpeningElement(path) {
+        const { node } = path
+        if (node.name.name !== imageImportName) {
+          return
+        }
+        const componentName = t.jsxIdentifier(`GatsbyImage`)
+
+        const fixedOrFluid = node.attributes.filter(
+          ({ name }) => name?.name === `fluid` || name?.name === `fixed`
+        )
+
+        const otherAttributes = node.attributes.filter(
+          ({ name }) => name?.name !== `fluid` && name?.name !== `fixed`
+        )
+
+        if (!fixedOrFluid.length > 0) {
+          path.replaceWith(
+            t.jsxOpeningElement(componentName, [...otherAttributes], true)
+          )
+          return
+        }
+        const expressionValue = fixedOrFluid?.[0]?.value?.expression
+
+        let newImageExpression = expressionValue // by default, pass what they pass
+
+        if (t.isMemberExpression(expressionValue)) {
+          const expressionStart =
+            expressionValue?.object?.object ?? expressionValue?.object
+
+          newImageExpression = template.expression
+            .ast`${expressionStart}.childImageSharp.gatsbyImageData`
+          newImageExpression.extra.parenthesized = false // the template adds parens and we don't want it to
+        } else if (t.isObjectExpression(expressionValue)) {
+          expressionValue.properties.map(item => {
+            if (item.key.name !== `src`) return
+            if (t.isMemberExpression(item.value)) {
+              let subObject = item.value?.object
+              while (subObject) {
+                if (
+                  subObject.property?.name === `fixed` ||
+                  subObject.property?.name === `fluid`
+                ) {
+                  subObject.property.name = `gatsbyImageData`
+                  break
+                } else {
+                  console.log(`WARN`)
+                }
+                subObject = subObject?.object
+              }
+            }
+          })
+        }
+
+        // // create new prop
+        const updatedAttribute = t.jsxAttribute(
+          t.jsxIdentifier(`image`),
+          t.jsxExpressionContainer(newImageExpression)
+        )
+
+        path.replaceWith(
+          t.jsxOpeningElement(
+            componentName,
+            [updatedAttribute, ...otherAttributes],
+            true
+          )
+        )
+        path.skip() // prevent us from revisiting these nodes
+      },
+      ClassDeclaration({ node }) {
+        if (node.superClass?.name === imageImportName) {
+          console.log(`WARN`)
+        }
```

Comment:
If this is to catch unsupported usages of `Img` in general, you could attach it to `ReferencedIdentifier`, and warn whenever the parent is not a JSX node.

Reply:
Couldn't find docs on that besides one example I couldn't abstract correctly. If you have another source I'd be happy to do that since that is the ideal case.